### PR TITLE
⬆️ Prepare upgrade handlers

### PR DESCRIPTION
**Pull request by bdeneux**
A PR to bring some functionalities to prepare upgrades of the chain.

- [x] Add an upgrade handler
- [x] Add migration script for chain
- [x] Add migration script for logic module
- [x] Add a make script to test chain upgrade
- [x] Add documentation for the make script
- [x] Increase the max gas limits

#### ⬆️ Upgrade handler 

Add necessary upgrade handlers to prepare chain upgrade and migration.

#### 🧠 Logic module

The logic module introduce new params and remove the foo/bar params. A migration has been configured to set the new params with its defaults values. Logic modules consensus version has been incremented to v2. A migration folder structure has been set with its related migration. 

#### ⚙️ Build

A new make command has been added to test chain upgrade : 
```bash
make chain-upgrade FROM_VERSION=v3.0.0 TO_VERSION=v4.0.0   
```

This command will build the current go project, download the old version from GitHub tag `FROM_VERSION`, build and start this version with cosmovisor. Vote proposal will be submitted and voted to make an upgrade after 10 block. Then the new binary is prepared to be updated when chain reach the block 10. Chain restart and migration name `TO_VERSION` is done (if necessary). 

#### 🔗 Resources 
Nice article that help to understand migration process in cosmos chain : https://medium.com/web3-surfers/cosmos-dev-series-cosmos-sdk-based-blockchain-upgrade-b5e99181554c

### CODE REVIEW

File: x/logic/types/params.go [hunk 1]
```diff
@@ -18,7 +18,7 @@ var (
 var (
 	DefaultRegisteredPredicates = make([]string, 0)
 	DefaultBootstrap            = ""
-	DefaultMaxGas               = math.NewUint(uint64(100))
+	DefaultMaxGas               = math.NewUint(uint64(10000))
```
Comment (amimart) [inline x/logic/types/params.go:21]: I'm wondering if it is sufficient, this is an open question, any remarks @ccamel ?

**Review by amimart — COMMENTED** [review]
Excellent! I just throw a question regarding gas limit default parameters, especially in the Nemeton phase 4 perspectives which will consist in testing the logic module.

File: x/logic/types/params.go (hunk 1 above)
Comment (amimart) [reply x/logic/types/params.go:21]: @ccamel @bdeneux I propose to set a limit of `100000` for testing purposes. We'll easily find out what we're capable of with it, and it'll help identify a sufficient and reasonable limit for mainnet.

Comment (bdeneux) [reply x/logic/types/params.go:21]:
```suggestion
	DefaultMaxGas               = math.NewUint(uint64(100000))
```
I'm agree with this value, I think it's a good start 👍

**Review by ccamel — APPROVED** [review]
Let's, go! 😎

**Review by amimart — APPROVED** [review]

**Pull request merged**